Add new parsers.py, unit tests (test_parsers.py) and resources.

mattjshannon opened the pull request.

mattjshannon committed:
Add new parsers.py, unit tests (test_parsers.py) and resources.

mattjshannon committed:
Remove pandas reference

mattjshannon committed:
Add lxml to requirements.txt

mattjshannon committed:
Typo on package_data

mattjshannon commented:
I'm going to add more unit tests to this PR, extending the code coverage.

mattjshannon committed:
Updated tests, includes experimental data now

PAHdb commented on .travis.yml:
```diff
@@ -26,7 +26,7 @@ script:
   - codecov  # Upload the report to codecov.
   - cd $TRAVIS_BUILD_DIR  # Change path back to $TRAVIS_BUILD_DIR.
-  - flake8 --exclude versioneer.py  # Enforce code style.
+  - flake8 --exclude versioneer.py,amespahdbpythonsuite/utils/token_utils.py  # Enforce code style.
```
token_utils is not needed for the parser alone--you can remove it here.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            try:
+                self.filename = os.environ["AMESPAHDEFAULTDB"]
+            except KeyError as e:
+                print("Must set AMESPAHDEFAULTDB environment variable.")
```
Message is incorrect; you can also pass a filename--please fix. I think it should also be possible to create an instance without providing a filename; something along the lines of the following should be possible:

import amespahdbpythonsuite.parser

p = parser()
p.filename = 'file.xml'
dict = p.to_dict(validate=True)

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            not (False). Defaults to False until validation is performed.
+    """
+
+    def __init__(self, filename=None, check_schema=True):
```
Do you think there should be a default init which would allow the following:

import amespahdbpythonsuite.parser
p = parser()
p.filename = 'file.xml'
dict = p.to_dict(validate=True)

or for the last step:
p.validate = True
dict = p.to_dict()

I don't know if it is the python-way to have non-parametric inits; C++ does, as you should be able to, say, make a vector of objects, which by default call the non-parametric constructor:

std::vector&lt;MyClass&gt; myVec(10);

mattjshannon replied:
Now supports this functionality.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+        """
+
+        if self.check_schema:
+            tree = ElementTree.parse(self.filename)
```
Should tree be a class-member, that way you don't have too pass it around as an argument.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+
+        if self.check_schema:
+            tree = ElementTree.parse(self.filename)
+            root = tree.getroot()
```
Should root be a class-member, that way you don't have too pass it around as an argument.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+        if self.check_schema:
+            tree = ElementTree.parse(self.filename)
+            root = tree.getroot()
+            self._validate_schema(tree, root)
```
With the above changes this would just become self.validate.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+                ElementTree.iterwalk(tree, events=("start", "end"))
+
+        else:
+            formatted_tree = \
```
Should formatted_tree be a class-member, that way you don't have too pass it around as an argument.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+"""
+parsers.py
+
+Parse a PAHdb XML file, with or without schema checking.
```
Perhaps you can inherit from etree and extend it with a 'to_dict'-method. That way you transparently offer the  etree API to users, but also 'our' interpretation. The 'to_dict'-method would also do the validation, if wanted.

To me this seems more OO-appropriate, as you are 'extending' the XML-functionality to generate a (non-general) dictionary.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+#!/usr/bin/env python3
+"""
+parsers.py
```
Perhaps for clarity the file should be named xmlparser.py.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+        self.check_schema = check_schema
+        self.validated = False  # Assume schema has not yet been checked.
+
+        # Identify location of PAHdb XML file.
```
This logic is probably better handled in amespahdbidlsuite.py, i.e., if the file doesn't exist lxml should throw an exception. The philosophy would be that "the parser only parses", i.e., has the to_dict-method.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+        """Inits XMLparser with schema checking on, no given filename."""
+        self.filename = filename
+        self.check_schema = check_schema
+        self.validated = False  # Assume schema has not yet been checked.
```
This belongs in amespahdbpythonsuite.py; you should be able to validate more than once; especially if the default lxml-api is exposed.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+    def __init__(self, filename=None, check_schema=True):
+        """Inits XMLparser with schema checking on, no given filename."""
+        self.filename = filename
+        self.check_schema = check_schema
```
This belongs in amespahdbpythonsuite.py.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+
+    def __str__(self):
+        """A description of the object."""
+        return (f'An XML parser from the Ames PAHdb Python Suite. '
```
It is AmesPAHdbPythonSuite.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            else:
+                print('Using default PAHdb file:\n{}'.format(self.filename))
+
+        # Parse XML file to dictionary.
```
This should be explicitly called by the user:

import amespahdbpythonsuite.parser

p = parser(Filename="file.xml")
p.validate() # a try here I suppose ...
dict = p.to_dict()

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+        return (f'An XML parser from the Ames PAHdb Python Suite. '
+                f'XML file: \n{self.filename!r}.')
+
+    def _parse(self):
```
This would be the 'to_dict'-method; lxml's parse would still be exposed as well. Perhaps 'to_dict' should be 'to_pahdb_dict' to make clear it is particular interpretation of the XML.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+                xmlschema.assertValid(tree)
+            except Exception as e:
+                raise e
+            else:
```
Should return True/False, not set a property, and not be private.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            else:
+                self.validated = True
+
+    def _tree_to_dict(self, tree):
```
Maybe rename to _tree_to_pahdb_dict?

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+                uid, metadata, comments, reference = \
+                    self._extract_info(elem, namespace)
+
+                # Extract hard data.
```
What is 'hard' data?

mattjshannon replied:
The numeric data -- updated the descriptions to be more accurate.

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+        full_path = os.path.abspath(self.filename)
+
+        # Sanity check, no repeated UIDs.
+        db_keys = [key for key in database_dict]
```
Is this not done by the schema-check; redundant here ...

mattjshannon replied:
I don't think this is done in the schema check currently, unless this is a general requirement of the XML schema?

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+
+        return database_dict, info_dict
+
+    def to_dicts(self):
```
Can be removed (?)

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+        """
+        namespace = ''
+        try:
+            namespace_end_index = root.tag.index('}')
```
Not very elegant, but I guess this is the only way ...

mattjshannon replied:
Not sure that it's easier on the eyes, but what about...
```suggestion
            namespace = root.tag.split('}')[0] + '}'
```

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            ref_list (tuple): Tuple of general references from XML file.
+        """
+
+        def get_metadata_keys():
```
Does this need to be a function, does a (class-level) variable not suffice?

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            `scale`, `intensity`, `symmetry`.
+        """
+
+        def geometry_keys():
```
Does this need to be a function, does a (class-level) variable not suffice?

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            }
+            return geo_keys
+
+        def transition_keys():
```
Does this need to be a function, does a (class-level) variable not suffice?

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+            }
+            return tran_keys
+
+        def laboratory_keys():
```
Does this need to be a function, does a (class-level) variable not suffice?

PAHdb commented on amespahdbpythonsuite/parsers.py:
```diff
@@ -0,0 +1,543 @@
+
+            for mode_property in mode_properties:
+                key = mode_property.tag.split(namespace)[-1]
+                value = mode_property.text  # of type str.
```
Type 'str' is not numeric; the method name is therefore misleading ...

mattjshannon replied:
The data is numeric but LXML returns the 'text' field (containing the data) as a string; shortly afterward it is cast to its appropriate numerical data type (3 lines below).

PAHdb commented on setup.py:
```diff
@@ -157,7 +157,7 @@
     include_package_data=True,
     package_data={  # Optional
-        'pypahdb': ['data/*'],
+        'amespahdbpythonsuite': ['resources/*'],
```
For pyPAHdb we had a data-directory; I like to keep things consistent between the different code-bases; either change here or in pyPAHdb---I do like calling it 'resources' better.

PAHdb commented on amespahdbpythonsuite/tests/test_parsers.py:
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+"""
+test_parsers.py
```
For pyPAHdb we didn't use pytest; I like to keep things consistent between the different code-bases; either change here or in pyPAHdb---pytest is probably the preferred way ...

PAHdb reviewed (changes requested):
As you will see, I have a few (re)design suggestions. Notably inheriting from lxml's etree and extending that with a 'to_pahdb_dict'-method. I also have a couple of more philosophical suggestions, including using class variables instead of passing arguments to methods within the class, consistency between pyPAHdb and the suite, and moving some logic to amespahdbidlsuite.py.

Is there somewhere I can look over the generated docs, perhaps you could run Sphinx locally and put a copy of the HTML-files in the Dropbox.

mattjshannon committed:
Incorporate CB comments from PR

API updated to support different calls. For instance, these are all valid:

        parser = XMLparser(filename="xml_file.xml")
        parser.verify_schema()
        database, info = parser.to_pahdb_dict()

        parser = XMLparser(filename="xml_file.xml", validate=True)
        database, info = parser.to_pahdb_dict()

        parser = XMLparser()
        parser.filename = "xml_file.xml"
        parser.validate = True
        database, info = parser.to_pahdb_dict()

Still need to add the from_string method code once the general class structure is OK.

mattjshannon committed:
Missing __init__.py I think

mattjshannon committed:
Flake8 fix

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+        return (f'An XML parser from the AmesPAHdbPythonSuite. '
+                f'XML file: \n{self.filename!r}.')
+
+    def verify_schema(self):
```
For clarity, maybe this method should be called 'check_with_linked_schema' or 'check_with_embedded_schema'. This way it is clear that it will use any linked schema.

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+        """Validate XML schema.
+
+        Note:
+            Requires that self.filename is set.
```
The method should throw an exception when self.filename is not set and/or when no linked schema is found.

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+            _root: root element of ElementTree tree.
+
+        Returns:
+            True if successful. Also assigns self.schema_is_valid to True
```
There should not be a self.schema_is_valid, as one could envision a scenario where the tree was updated.

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+
+        if schema:
+            _, uri = schema.split(' ', 1)
+            doc = ElementTree.parse(urllib.request.urlopen(uri))
```
I presume urllib and ElementTree.parse will either throw an exception when the XML-document doesn't contain a 'SchemaLocation' or when the 'parse' fails?

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+            overridden.
+
+        Note:
+            Sets the attributes self.database and self.info if successful.
```
I don't think this information should be kept as class properties, as the filename/XML-string can be updated outside this method.

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+        events = ("start", "end")
+
+        # First determine whether we've already validated the schema.
+        if self.schema_is_valid:
```
self.schema_is_valid should be removed. If somebody would call this method and wants validation, it just will have to parse the XML again. That is if someone would do:

1.        parser = XMLparser()
2.       parser.filename = "xml_file.xml"
3.        parser.check_schema()
4.        parser.validate = True
5.       database, info = parser.to_pahdb_dict()

the validation would happen twice, i.e., in line 3 and 5.

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+    @classmethod
+    def from_string(cls, string):
+        """Generate XMLparser instance from a string."""
+        # TODO: complete this method, if the general class structure is OK.
```
OK!

PAHdb commented on amespahdbpythonsuite/xmlparser.py:
```diff
@@ -0,0 +1,525 @@
+            per atom in the geometry or vibrational mode is present in
+            the tuple (respectively).
+
+        Note:
```
I am a bit unsure on what the best way is to connect specific 'tag-names' with their processing logic. On the one hand, I do like having the tag-names in 'utils/xml_data_types.py'. On the other hand, there seems to be a disconnect, i.e., the XMLParser is not 'standalone/self-contained'.

I've seen examples where the parsing is entirely event-driven, which I suspect is best-practice with the SAP-like-API. That is, instead of explicitly calling functions upon receiving a tag-name-event,  you have similar to what is in 'xml_data_types.py' a dictionary with the tag-name as key and the value a function to be called. Obviously there would be a limit to how deep you would go, e.g., calling a separate function for each 'frequency'-tag would be too much. On the other hand, higher-level tags like 'specie', 'geometry', 'transitions', etc. would be perfectly suited for this approach. Each function would have a only a few if-then-else-statements to recognize the event--the advantage is that events are handled as soon as they come in and not, as now, an enclosing tag has come in. For example, the first level of the iterwalk/parse would simply handle 'comment' and 'species'. On receiving the 'species'-opening tag a function 'parse_species' is called which takes the tree and continues parsing by starting with handling the 'specie'-tags by calling the 'parse_specie' function , which checks handles the 'geometry', 'transitions', etc. tags through appropriate function calls. The parsed data would be stored in a growing dictionary. The species properties could be handled in the 'parse_specie'-function itself. In all cases, when a tag-name is not recognized/in the tag-name/function-dictionary an exception is thrown. In my opinion, the processing/parsing should be more like a 'waterfall'/'tree' as that mimics the structure of the data and the API used. Thinking about adapting the parser for the forthcoming cluster data, this would the be relatively straight-forward, as the only thing that needs to happen is to add an entry in the tag-name/function-dictionary to the appropriate function handling that tag. Note that the only thing the parser in essence does is translating repeated tags into lists of dictionaries and casting the data from string into the proper type.

What are your thoughts?

PAHdb reviewed (changes requested):
Hi Matt,

Overall good improvements. I think simplification is key here. Below somewhere I have a somewhat long philosophical rant on how I think we could maybe bring some more coherence to the code. Of course, the code currently works as intended and it is hard to envision anybody using it any other way than that. However, I do think it is a good exercise to go through it this way as it will help when things get more complicated.

PAHdb commented:
c7fdf33f84356ec589a90e644e33cdd2592168b5 implements the changes of the pull request.
